cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java

```diff
@@ -104,20 +120,144 @@ private boolean upgrade(Id.Application appId) throws Exception {
+      // "adapterDrafts": {
+      //   "default": {
+      //     "TwitterToHBase": { ... }
+      //   },
+      //   ...,
+      //   "isMigrated": true
+      // }
+      //
+      // Note the 'isMigrated' at the bottom which is a boolean and not an object like all the other keys...
+      Map<String, Map<String, PipelineDraft>> drafts = new HashMap<>();
+      try {
+        // try to deserialize with 3.2 format
+        Map<String, PipelineDraft> defaultDrafts =
+          GSON.fromJson(consoleData.get("adapterDrafts"),
+                        new TypeToken<Map<String, PipelineDraft>>() { }.getType());
+        drafts.put("default", defaultDrafts);
+      } catch (JsonSyntaxException e) {
+        // try to deserialize with 3.3 format
+        // need to remove 'isMigrated' otherwise this will not work, since a boolean is not a map.
+        JsonObject draftsJson = consoleData.get("adapterDrafts").getAsJsonObject();
+        draftsJson.remove("isMigrated");
+        drafts = GSON.fromJson(consoleData.get("adapterDrafts"),
+                               new TypeToken<Map<String, Map<String, PipelineDraft>>>() { }.getType());
+      }
+
+      for (Map.Entry<String, Map<String, PipelineDraft>> namespaceEntry : drafts.entrySet()) {
+        String namespace = namespaceEntry.getKey();
+
+        Map<String, PipelineDraft> namespaceDrafts = new HashMap<>();
+        for (Map.Entry<String, PipelineDraft> draft : namespaceEntry.getValue().entrySet()) {
+          String draftName = draft.getKey();
+          PipelineDraft draftSpec = draft.getValue();
+          if (draftSpec.isOldBatch()) {
+            System.out.println("Upgrading draft " + draftName + " in namespace " + namespace);
+            PipelineDraft upgradedDraft = draftSpec.getUpgradedBatch(batchArtifact, batchPluginArtifactFinder);
+            upgradedDraft.setName(draftName);
+            namespaceDrafts.put(draftName, upgradedDraft);
+          } else if (draftSpec.isOldRealtime()) {
+            System.out.println("Upgrading draft " + draftName + " in namespace " + namespace);
+            PipelineDraft upgradedDraft = draftSpec.getUpgradedRealtime(realtimeArtifact, realtimePluginArtifactFinder);
+            upgradedDraft.setName(draftName);
+            namespaceDrafts.put(draftName, upgradedDraft);
+          } else {
+            namespaceDrafts.put(draftName, draftSpec);
+          }
+        }
+        drafts.put(namespace, namespaceDrafts);
+      }
+      consoleData.remove("adapterDrafts");
+      // writing to 'hydratorDrafts' instead of 'adapterDrafts' now
+      consoleData.add("hydratorDrafts", GSON.toJsonTree(drafts));
+      // ... without this, the UI will move everything in hydratorDrafts one level down under 'default'.
+      // but what happens if the next version requires migration?
+      consoleData.getAsJsonObject("hydratorDrafts").addProperty("isMigrated", true);
+      upgradedSomething = true;
+    }
+
+    if (consoleData.has("pluginTemplates")) {
```

Comment:
nit: should we extract these keys (even as local vars since they are used multiple times)